data-model/src/main/java/org/hypertrace/core/datamodel/shared/StructuredTraceGraph.java

```diff
@@ -20,187 +8,41 @@
-        Event childEvent = events.get(targetIndex);
-        parentToChildrenEvents.computeIfAbsent(parentEvent.getEventId(), k -> new ArrayList<>())
-            .add(childEvent);
-        childToParentEvents.put(childEvent.getEventId(), parentEvent);
-      }
-    }
-
-    // return for now, we don't forsee this is to be null.
-    if (entities == null) {
-      return;
+  public static StructuredTraceGraph reCreateTraceEventsGraph(StructuredTrace trace) {
+    if (null == graph) {
+      return createGraph(trace);
     }
-
-    List<Edge> entityEdges = trace.getEntityEdgeList();
-    if (entityEdges != null) {
-      for (Edge entityEdge : trace.getEntityEdgeList()) {
-        Integer sourceIndex = entityEdge.getSrcIndex();
-        Integer targetIndex = entityEdge.getTgtIndex();
-        Entity parentEntity = entities.get(sourceIndex);
-        Entity childEntity = entities.get(targetIndex);
-        parentToChildrenEntities.computeIfAbsent(parentEntity.getEntityId(), k -> new ArrayList<>())
-            .add(childEntity);
-        childToParentEntities.computeIfAbsent(childEntity.getEntityId(), k -> new ArrayList<>())
-            .add(parentEntity);
-      }
-    }
-  }
-
-  private void buildRootEvents(StructuredTrace trace) {
-    // build the root event ids
-    Set<ByteBuffer> rootEventIds = trace.getEventList().stream().map(Event::getEventId)
-        .collect(Collectors.toSet());
-
-    // remove all the children, and what's remaining are the events without children
-    // we will consider these are roots, including the ones that are standalone
-    Set<ByteBuffer> childrenEventIds = parentToChildrenEvents.values().stream()
-        .flatMap(l -> l.stream().map(Event::getEventId)).collect(Collectors.toSet());
-    rootEventIds.removeAll(childrenEventIds);
-
-    rootEvents = rootEventIds.stream().map(eventMap::get).collect(Collectors.toUnmodifiableSet());
-  }
-
-  private void buildRootEntities(StructuredTrace trace) {
-    // build the root entity ids
-    Set<String> rootEntityIds = trace.getEntityList().stream().map(Entity::getEntityId)
-        .collect(Collectors.toSet());
-
-    // remove all the children, and what's remaining are the entities without children
-    // we will consider these are roots, including the ones that are standalone
-    Set<String> childrenEntityIds = parentToChildrenEntities.values().stream()
-        .flatMap(l -> l.stream().map(Entity::getEntityId)).collect(Collectors.toSet());
-    rootEntityIds.removeAll(childrenEntityIds);
-
-    rootEntities = rootEntityIds.stream().map(entityMap::get)
-        .collect(Collectors.toUnmodifiableSet());
+    graph.traceEventsGraph = TraceEventsGraph.createGraph(trace);
+    return graph;
   }
 
-  /**
-   * @return an immutable map of event ids to events
-   */
-  public Map<ByteBuffer, Event> getEventMap() {
-    return eventMap;
-  }
+  public static StructuredTraceGraph reCreateTraceEntitiesGraph(StructuredTrace trace) {
```

Comment:
How are we planning to use this downstream in ingester with thread-local - StructuredTraceGraphBuilder?
So, we don't want to build an event graph every time the structure trace has changed due to a newly added/updated entity?

Reply:
So, I think, we will check if trace has modified for entity or event, and accordingly reCreate the state, right?

Reply:
yes we check which one of Event/Entities graph needs rebuilding